src/go/rdctl/cmd/snapshot.go

```diff
@@ -63,38 +63,46 @@ func exitWithJsonOrErrorCondition(e error) error {
 // If the main process is running, stops the backend, calls the
 // passed function, and restarts the backend. If it cannot connect
 // to the main process, just calls the passed function.
-func wrapSnapshotOperation(wrappedFunction cobraFunc) cobraFunc {
-	return func(cmd *cobra.Command, args []string) error {
-		appPaths, err := paths.GetPaths()
-		if err != nil {
-			return fmt.Errorf("failed to get paths: %w", err)
-		}
-		if err := createBackendLock(appPaths.AppHome); err != nil {
-			return err
-		}
-		defer removeBackendLock(appPaths.AppHome)
-		if err := ensureBackendStopped(cmd); err != nil {
-			return err
-		}
-		if err := wrappedFunction(cmd, args); err != nil {
+func wrapSnapshotOperation(cmd *cobra.Command, appPaths paths.Paths, resetOnFailure bool, wrappedFunction cobraFunc) error {
+	if err := createBackendLock(appPaths.AppHome); err != nil {
+		return err
+	}
+	defer removeBackendLock(appPaths.AppHome)
+	if err := ensureBackendStopped(cmd); err != nil {
+		return err
+	}
+	if err := wrappedFunction(); err != nil {
+		if resetOnFailure {
 			factoryreset.DeleteData(appPaths, true)
-			return err
 		}
-		// Note that this does not wait for the backend to be in the
-		// STARTED (or DISABLED if k8s is disabled) state. This allows
-		// removeBackendLock() to be called as a deferred function while
-		// keeping the state of the backend lock file in sync with the
-		// main process backendIsLocked variable.
-		return ensureBackendStarted()
+		return err
+	}
+	// Note that this does not wait for the backend to be in the
+	// STARTED (or DISABLED if k8s is disabled) state. This allows
+	// removeBackendLock() to be called as a deferred function while
+	// keeping the state of the backend lock file in sync with the
+	// main process backendIsLocked variable.
```

Comment:
I don't understand what this means.

The "keep in sync" explanation worries me: there should be only a single place that the state is kept (in the file), and the code in `rdctl` should simply notify the app that it needs to recheck the file because the status may have changed.

Reply:
This is an existing comment that got moved around, not part of this PR and not something I wrote.

Reply:
I agree that the way it works right now is not ideal. But this is the way it works right now (and it does work), and we are trying to release. Can we deal with this after 1.11?